repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/apivalidator/ApiValidatorMultiTenantTest.groovy

```diff
@@ -0,0 +1,475 @@
+
+        def params = properties.getDefaultTemplateParams()
+        repose.configurationProvider.applyConfigs("common", params)
+        repose.configurationProvider.applyConfigs("features/filters/apivalidator/multitenant", params)
+        repose.start()
+
+        repose.waitForNon500FromUrl(reposeEndpoint)
+    }
+
+    @Unroll("Should succeed on GET /v1/resource/other when appropriate headers are set (#tenants)")
+    def "Should succeed when appropriate headers are set"() {
+        when:
+        MessageChain mc = deproxy.makeRequest(
+            url: "$reposeEndpoint/v1/resource/other",
+            method: "GET",
+            headers: [
+                "X-Tenant"   : tenants,
+                "X-Other"    : "other",
+                (ROLES)      : "foo,bar",
+                "X-Map-Roles": mapHeaderValue])
+
+        then:
+        mc.getReceivedResponse().getCode() as Integer == SC_OK
+        mc.handlings.size() == 1
+        List<String> relevantRolesValues = mc.handlings[0].request.headers.findAll(RELEVANT_ROLES)
+        relevantRolesValues.containsAll(["foo", "bar"])
+
+        where:
+        tenants << ["1,5", "4", "4,2", "1,2,3", "5,2", "1,3", "5,3", "1", "1,5,7", "1,5,6"]
+    }
+
+    @Unroll("Should succeed on #method #url when appropriate headers are set (1, 5)")
+    def "Should succeed when appropriate headers are set (1, 5)"() {
+        when:
+        MessageChain mc = deproxy.makeRequest(
+            url: "$reposeEndpoint$url",
+            method: method,
+            headers: [
+                "X-Tenant"   : "1,5",
+                "X-Other"    : "other",
+                (ROLES)      : "foo,bar",
+                "X-Map-Roles": mapHeaderValue])
+
+        then:
+        mc.getReceivedResponse().getCode() as Integer == SC_OK
+        mc.handlings.size() == 1
+        List<String> relevantRolesValues = mc.handlings[0].request.headers.findAll(RELEVANT_ROLES)
+        relevantRolesValues.contains("a:admin/{X-TENANT}")
+
+        where:
+        [method, url] << observerRequests + creatorRequests + updaterRequests + adminOnlyRequests
+    }
+
+    def "Should succeed on GET /v1/resource/other when appropriate headers are set (1,5)"() {
+        when:
+        MessageChain mc = deproxy.makeRequest(
+            url: "$reposeEndpoint/v1/resource/other",
+            method: "GET",
+            headers: [
+                "X-Tenant"   : "1,5",
+                "X-Other"    : "other",
+                "X-Map-Roles": mapHeaderValue])
+
+        then:
+        mc.getReceivedResponse().getCode() as Integer == SC_OK
+        mc.handlings.size() == 1
+        List<String> relevantRolesValues = mc.handlings[0].request.headers.findAll(RELEVANT_ROLES)
+        relevantRolesValues.isEmpty()
+    }
+
+    @Unroll("Should fail on #method #url when appropriate headers are set but there's no tenant access (#tenants)")
+    def "Should fail on when appropriate headers are set but there's no tenant access"() {
```

Comment:
"should fail on when" should probably just be "should fail when". Not that it would ever matter...

Reply:
Yeah, I was just removing the variable parts of the `Unroll` name for the `def` name.

Reply:
I thought as much. Not a big deal, but eh.

Reply:
This was stupid naming anyway. I don't know why I did it. They are now fixed.